src/js/tools/sentryMetroSerializer.ts

```diff
@@ -0,0 +1,256 @@
+ */
+const createDefaultMetroSerializer = (serializerConfig: Partial<SerializerConfigT>): MetroSerializer => {
+  return (entryPoint, preModules, graph, options) => {
+    const createModuleId = options.createModuleId;
+    for (const module of graph.dependencies.values()) {
+      options.createModuleId(module.path);
+    }
+    const modules = [...graph.dependencies.values()].sort((a, b) => createModuleId(a.path) - createModuleId(b.path));
+
+    let bundle = baseJSBundle(entryPoint, preModules, graph, options);
+    if (options.sentryBundleCallback) {
+      bundle = options.sentryBundleCallback(bundle);
+    }
+    const { code } = bundleToString(bundle);
+    if (serializerConfig.processModuleFilter === undefined) {
+      // processModuleFilter is undefined when processing build request from the dev server
+      return code;
+    }
+
+    // Always generate source maps, can't use Sentry without source maps
+    const map = sourceMapString([...preModules, ...modules], {
+      processModuleFilter: serializerConfig.processModuleFilter,
+      shouldAddToIgnoreList: options.shouldAddToIgnoreList,
+    });
+    return { code, map };
+  };
+};
+
+/**
+ * Looks for a particular string pattern (`sdbid-[debug ID]`) in the bundle
+ * source and extracts the bundle's debug ID from it.
+ *
+ * The string pattern is injected via the debug ID injection snipped.
+ *
+ * https://github.com/getsentry/sentry-javascript-bundler-plugins/blob/40f918458ed449d8b3eabaf64d13c08218213f65/packages/bundler-plugin-core/src/debug-id-upload.ts#L293-L294
+ */
+function determineDebugIdFromBundleSource(code: string): string | undefined {
+  const match = code.match(
+    /sentry-dbid-([0-9a-fA-F]{8}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{12})/,
+  );
+
+  if (match) {
+    return match[1];
+  } else {
+    return undefined;
+  }
+}
+
+/**
+ * Creates a Metro serializer that adds Debug ID module to the plain bundle.
+ * The Debug ID module is a virtual module that provides a debug ID in runtime.
+ *
+ * RAM Bundles do not support custom serializers.
+ */
+export const createSentryMetroSerializer = (customSerializer?: MetroSerializer): MetroSerializer => {
+  return async function (this: ExpectedSerializedConfigThisContext, entryPoint, preModules, graph, options) {
+    let currentDebugIdModule: Module<MixedOutput> | undefined;
+    const serializer = customSerializer || createDefaultMetroSerializer(this);
+    options.sentryBundleCallback = sentryBundleCallback;
+
+    // Add debug id module to the preModules
+    let modifiedPreModules: readonly Module<MixedOutput>[];
+    const containsDebugIdModule = preModules.some(module => module.path === DEBUG_ID_MODULE_PATH);
+    if (!containsDebugIdModule) {
+      const debugIdModule = createDebugIdModule(DEBUG_ID_PLACE_HOLDER);
+      const tmpPreModules = [...preModules];
+      if (tmpPreModules.length > 0 && tmpPreModules[0] !== undefined && tmpPreModules[0].path === PRELUDE_MODULE_PATH) {
```

Comment:
m: I'd like it if both the logic to generate `modifiedPreModules` and `bundleCode`/`bundleMapString` was extracted into helper functions so there is a clear input/output.